### ISSUE TITLE
Wrong exception thrown when using `with_job` decorator.

### ISSUE THREAD

**Issue by kidrahahjo** [glotzerlab/signac-flow#600]
### Description
A `FlowProjectDefinitionError` with the message _"The @with_job decorator cannot be used with aggregation"_ is thrown even when we don't use the `aggregator` decorator.

The error is caused due to ordering of the decorators. If we add it after the `FlowProject.operation` decorator, then it behaves as expected.

### To reproduce
Script:
```python
# project.py
from flow import FlowProject, with_job

@with_job
@FlowProject.operation
def with_job_op(job):
    print("with_job_op")

if __name__ == "__main__":
    FlowProject().main()
```

Run the above script with 
```bash
python project.py run
```

### Error output

> Traceback (most recent call last):
  File "workflow.py", line 36, in <module>
    def with_job_op(job):
  File "/c/opensource/signac-flow/flow/operations.py", line 99, in with_job
    raise FlowProjectDefinitionError(
flow.errors.FlowProjectDefinitionError: The @with_job decorator cannot be used with aggregation.
### System configuration

Please complete the following information:

 - Operating System [e.g. macOS]: Linux-5.10.16.3-microsoft-standard-WSL2-x86_64-with-glibc2.29
 - Version of Python [e.g. 3.7]: 3.8.10 (default, Nov 26 2021, 20:14:08), [GCC 9.3.0]
 - Version of signac [e.g. 1.0]: 1.7.0
 - Version of signac-flow: 0.17.0

Or copy & paste the output of: `python -c 'import platform; print(platform.platform()); import sys; print(sys.version); import signac; print(signac.__version__); import flow; print(flow.__version__)'`

**Comment by kidrahahjo** [comment]
Not sure if we have prioritised the ordering requirements of the decorators, but I think, in future, adding more decorator friendly functionality to the package will make it difficult for us to maintain.

**Comment by bdice** [comment]
We could probably rewrite the error message to say, "The @with_job decorator can only be used on FlowProject operations, and does not support aggregation."

**Comment by vyasr** [comment]
> Oh, right. The error shouldn't show up at all. If this worked before we added that error for aggregation, then we should try to fix it as a regression. If that's too hard, we can require an order for the decorators and prohibit applying `@with_job` before declaring the function is a `@FlowProject.operation`.

While this solution would work here, I think @kidrahahjo is right to say that 

> Not sure if we have prioritised the ordering requirements of the decorators, but I think, in future, adding more decorator friendly functionality to the package will make it difficult for us to maintain.

We may want to explore alternate syntaxes that have less room for ambiguity. Specifically, I'm thinking that the API would be cleaner if we put more functionality into the `OperationsRegister` class instead of using so many decorators. The `with_job` and `cmd` decorators really could just be part of the call operator. For example, how would we feel about moving towards an API that looks more like
```
@Project.operation("op", with_job=True, cmd=True).with_directives(...)
def operation(job):
    return "echo `pwd`"
```
Groups and aggregations could also be added to the call operator (e.g. `Project.operation(..., groups=[g1, g2], aggregator=agg)`), or we could chain additional methods like `with_directives` (e.g. `Project.operation(...).groups([g1, g2]).aggregator(agg)`). Is there a benefit to the decorator-based API over this (other than backwards compatibility)?

CC @b-butler @csadorf @atravitz

**Comment by b-butler** [comment]
I think linking more of the functionality with the `operation` decorator makes sense. As to the exact form of that I am not sure. It depends on much we are willing to deprecate and eventually change the style. I am personally fine with deprecating `with_job` and `cmd`. If we go with @vyasr idea we would need to deprecate the current form of `operation.with_directives` which is fine we did introduce it more recently so it likely isn't being used very frequently.

**Comment by vyasr** [comment]
Why would we need to deprecate `with_directives`? I wasn't suggesting that, in fact my suggestion was intended to lean in to the preference for `@operation.with_directives` over `@directives`. Is my suggestion incompatible with that? Are you concerned about the ambiguity of whether directives assigned via `@operation(..., groups=[g1])).with_directives(...)` would apply to the operation or the group? To be clear, I think we could make the changes to `with_job` and `cmd` without changing the way groups and aggregators work if we think the additional complexity of the latter merits further discussion on the appropriate APIs.

**Comment by vyasr** [comment]
Ah I see the issue, I forgot that `with_directives` is implemented as a class method. Yeah, there would potentially be some breakage there. It may be possible to put in backwards compatibility layers to avoid breaking users by enabling multiple different patterns, but it might be too much maintenance burden. I'd say we should evaluate and focus on whichever changes we can make with the least disruption to existing workflows.

**Comment by kidrahahjo** [comment]
@vyasr Can this be a good small project for GSoC or do you think it is a more complicated design problem that needs a better understanding?

**Comment by vyasr** [comment]
@kidrahahjo good find. I think this change would be reasonable in scope for a "small" GSoC project if we could scope it correctly, but I can't be sure without a bit more of a concrete proposal. Could you compile the current set of decorators that we have and try to write up a spec of what we'd like an ideal API to look like, as well as what backwards compatibility layers that we might want to support? If you can add that to our doc then we can evaluate from there.

**Comment by b-butler** [comment]
This would. Although, I might just move to something like you suggested. The question comes in do we want a separate `with_directives` decorator when we could just

```python
@FlowProject.operation(
    name="foo",
    with_job=True,
    directives={"np": 4}
)
def operation_foo(job):
    pass
```

**Comment by b-butler** [comment]
Fixed in #686.
